### DISCUSSION TITLE
/ (which is left-associative) and /: (which is right-associative) have same precedence and may not be mixed

### DISCUSSION THREAD

**Discussion by jan0sch** [http4s/http4s#6595] [Q&A]
Hi, I'm having trouble with mixing path parameters and capturing the rest of the path:

```
case authedRequest @ GET -> Root / MyParamA(myA) / MyParamB(b) /: restOfThePath as user => ???
```

I get the following compiler error (using http4s 1.0.0-M35 and Scala 3.1.3):

```
/ (which is left-associative) and /: (which is right-associative) have same precedence and may not be mixed
```

Any ideas how this could be tackled? I would like to avoid something like `GET -> Root /: rest` and then fiddle everything out in custom code resulting in hand crafted redirects and so on.

**Comment by armanbilge** [comment]
Thanks for reporting. This seems annoying and maybe should even be an issue/bug.

If you add parentheses does that help? e.g. (but maybe try a few combinations):

```scala
case authedRequest @ GET -> (((Root / MyParamA(myA)) / MyParamB(b)) /: restOfThePath) as user => ???
```

or maybe even
```scala
case authedRequest @ GET -> Root./(MyParamA(myA))./(MyParamB(b))./:(restOfThePath) as user => ???
```

**Reply by jan0sch** [reply]
Thanks, it seems that parentheses do the trick: `GET -> (Root / MyParamA(myA) / MyParamB(myB)) /: rest => ???` compiles fine and I get a `rest: Path` which I can handle in the route. :-)

**Reply by jan0sch** [reply]
Okay, weird. It worked shortly but after a `sbt clean` it doesn't compile any longer. Instead I get the following error:

```
[error] 240 |    case ar @ GET -> (Root / MyParamA(myA) / MyParamB(myB)) /: rest
[error]     |                      ^
[error]     |this case is unreachable since type String is not a subclass of class Path
```

I'll play a bit more with parentheses and will report back here.

**Reply by jan0sch** [reply]
Played a bit and it seems nothing goes here. Even simple things like `GET -> (Root / "a") /: rest => ???` lead to `this case is unreachable since type String is not a subclass of class Path` errors.

Here are some things tried, you can exchange any of the `MyParamX(myX)` parts with static strings and will get the same errors nonetheless.

`GET -> MyParamA(myA) /: rest => ???` -> works

`GET -> Root / MyParamA(myA) /: rest => ???` -> `/ (which is left-associative) and /: (which is right-associative) have same precedence and may not be mixed`

`GET -> (Root / MyParamA(myA)) /: rest => ???` -> `this case is unreachable since type String is not a subclass of class Path`

`GET -> MyParamA(myA) / MyParamB(myB) /: rest => ???` -> `/ (which is left-associative) and /: (which is right-associative) have same precedence and may not be mixed`

`GET -> (MyParamA(myA) / MyParamB(myB)) /: rest => ???` -> `this case is unreachable since type String is not a subclass of class Path`

`GET -> (Root / MyParamA(myA) / MyParamB(myB)) /: rest => ???` -> `this case is unreachable since type String is not a subclass of class Path`

**Reply by rossabaker** [reply]
Not by a computer, but try `GET -> MyParamA(myA) /: MyParamB(myB) /: rest => ???`.  Once you go right-associative, you have to stay right-associative, and drop the `Root`.

**Reply by jan0sch** [reply]
Thank you very much Ross, that works. :heart: 

Should I drop a PR with a note regarding that into the docs?

**Reply by armanbilge** [reply]
@jan0sch yes please a PR would be great!

**Reply by jan0sch** [reply]
Here you go: https://github.com/http4s/http4s/pull/6598
:-)
